### PR TITLE
AK+LibJS: Exact and fast floating point to string conversion

### PR DESCRIPTION
**Pull request by DanShaders**
Currently, the floating point to string conversion is implemented several times across the codebase. Furthermore, there are some test failures in LibJS because of multiple imprecise algorithms implemented there. So, this PR aims to provide a single, fast and precise algorithm of doing the aforementioned conversion.

The algorithm is Ryu from https://dl.acm.org/doi/pdf/10.1145/3192366.3192369, which does not require any arbitrary precision arithmetic.

This is WIP but, as it is my first contribution, I want some comments on the already added code.

### CODE REVIEW

File: AK/StringFloatingPointConversions.h [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+ * Copyright (c) 2022, Dan Klishch <danilklishch@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#ifdef KERNEL
+#    error This file should not be included in the KERNEL as it deals with doubles \
+           and there is no guraantee does not do any floating point computations.
```
Comment (Hendiadyoin1) [inline AK/StringFloatingPointConversions.h:11]: missing "it"

File: AK/StringFloatingPointConversions.cpp [hunk 2]
```diff
@@ -0,0 +1,2279 @@
...
+    print(f"\t{{{q}, {k}, {{{coefficient & ((1 << 64) - 1)}ull, {coefficient >> 64}ull}}}},")
+
+*/
+// FIXME: Should be possible to compress from the current (48Kb) size to the one
+//        mentioned in the article (~9.5Kb).
+struct {
+    i32 q;
+    i32 k;
+    u64 multiplier[2];
+} DOUBLE_LOOKUP_TABLE[] = {
```
Comment (Hendiadyoin1) [inline AK/StringFloatingPointConversions.cpp:88]: constants still use snake case in our coding style,
same in the implementation functions

File: AK/StringFloatingPointConversions.cpp [hunk 3]
```diff
@@ -0,0 +1,2279 @@
...
+    [mn, mx] = minmax_euclid(a % b, b, maxw)
+
+    k = ceil(log2(maxw * a * b / (b - mx)))
+    coefficient = 2 ** k // 5 ** q + 1
+
+    print(f"\t{{{q}, {k}, {{{coefficient & ((1 << 64) - 1)}ull, {coefficient >> 64}ull}}}},")
+
+*/
+// FIXME: Should be possible to compress from the current (48Kb) size to the one
+//        mentioned in the article (~9.5Kb).
```
Comment (Hendiadyoin1) [inline AK/StringFloatingPointConversions.cpp:83]: one thing is that `q` and `k` seem to be fitting into `i16`s

File: AK/StringFloatingPointConversions.cpp [hunk 4]
```diff
@@ -0,0 +1,2279 @@
...
+    { 289, 732, { 2247116418577894885ull, 0ull } },
+    { 290, 732, { 449423283715578977ull, 0ull } },
+    { 290, 733, { 898846567431157954ull, 0ull } },
+    { 290, 734, { 1797693134862315908ull, 0ull } },
+};
+
+// Checks if value is divisible by 5 ^ power.
+bool is_divisible_by_pow_5(u64 value, i32 power)
+{
+    static const u64 FIVE_POWERS[] = { 5, 25, 625, 390625, 152587890625 };
```
Comment (Hendiadyoin1) [inline AK/StringFloatingPointConversions.cpp]:
```suggestion
    constepxr Array<u64> powers_of_five[] = { 5, 25, 625, 390625, 152587890625 };
```

File: AK/StringFloatingPointConversions.cpp [hunk 5]
```diff
@@ -0,0 +1,2279 @@
...
+{
+    static const u64 LEN_M = 52;
+    static const i32 BIAS = 1023;
+    static const i32 LOOKUP_OFFSET = BIAS + LEN_M + 1;
+
+    u64 bit_representation = bit_cast<u64, double>(value);
+
+    bool sign = bit_representation >> 63;
+    i32 e = static_cast<i32>(bit_representation >> LEN_M) & ((1 << 11) - 1);
+    u64 m = bit_representation & ((1ull << LEN_M) - 1);
```
Comment (Hendiadyoin1) [inline AK/StringFloatingPointConversions.cpp:2148]: how about a bitfield union, like LibC/math.cpp uses:
https://github.com/SerenityOS/serenity/blob/db2e1bfa027774ff1e1f883c4ddadc277c1e6e1e/Userland/Libraries/LibC/math.cpp#L85-L98

File: AK/StringFloatingPointConversions.cpp [hunk 6]
```diff
@@ -0,0 +1,2279 @@
...
+    u64 w = 4 * mf + 2;
+    // u * 2 ^ e2 < abs(answer) < w * 2 ^ e2        (1)
+    // abs(value) = v * 2 ^ e2 (yet another representation)
+
+    // Computes `operand * multiplier >> shift_amount'
+    auto multiply_and_shift = [&](u64 operand, u64 multiplier[2], i32 shift) {
+        using u128 = unsigned __int128;
+
+        u128 a = static_cast<u128>(operand) * multiplier[0];
+        u128 b = static_cast<u128>(operand) * multiplier[1] + (a >> 64);
```
Comment (Hendiadyoin1) [inline AK/StringFloatingPointConversions.cpp:2180]:
```suggestion
        u128 b = static_cast<u128>(operand) * multiplier[1] + a.high();
```
maybe `a.low()` in the lines below instead of casting to `u64`, but both are equivalent

File: AK/StringFloatingPointConversions.cpp [hunk 7]
```diff
@@ -0,0 +1,2279 @@
...
+        u64 c = static_cast<u64>(a);
+
+        if (0 <= shift && shift < 64) {
+            return (c >> shift) | static_cast<u64>(b << (64 - shift));
+        } else if (shift < 0) {
+            return c << (-shift);
+        } else {
+            assert(shift >= 64);
```
Comment (Hendiadyoin1) [inline AK/StringFloatingPointConversions.cpp:2188]:
```suggestion
            VERIFY(shift >= 64);
```

File: AK/StringFloatingPointConversions.cpp [hunk 8]
```diff
@@ -0,0 +1,2278 @@
...
+    }
+
+    // Step 4’. Find the shortest, correctly-rounded decimal representation in the interval.
+    bool is_even = ~m & 1;
+    bool accept_smaller = is_even && za;
+    bool accept_larger = is_even || !zc;
+
+    // a, b, c = u, v, w
+
+    if (!accept_larger) {
```
Comment (Hendiadyoin1) [inline AK/StringFloatingPointConversions.cpp:2241]: no braces are needed here

File: AK/StringFloatingPointConversions.cpp (hunk 3 above)
Comment (DanShaders) [reply AK/StringFloatingPointConversions.cpp:83]: This seems not to reduce size because of u64 aligment to 8 bytes

File: AK/StringFloatingPointConversions.cpp (hunk 4 above)
Comment (DanShaders) [reply AK/StringFloatingPointConversions.cpp]: Should I do the same with double_lookup_table?

File: AK/StringFloatingPointConversions.cpp (hunk 5 above)
Comment (davidot) [reply AK/StringFloatingPointConversions.cpp:2148]: I agree, also we should probably put that somewhere multiple spots can use it, I think (Un)SignedBitInteger also has a copy already

File: AK/StringFloatingPointConversions.cpp [hunk 9]
```diff
@@ -0,0 +1,2278 @@
...
+
+#include <AK/BitCast.h>
+#include <AK/BuiltinWrappers.h>
+#include <AK/StringFloatingPointConversions.h>
+
+namespace AK {
+
+// This entire algorithm is an implementation of the paper: Ryu: Fast Float-to-String Conversion
+// by Ulf Adams, available at https://dl.acm.org/doi/pdf/10.1145/3192366.3192369 and an implemenetation
+// at https://github.com/ulfjack/ryu.
```
Comment (davidot) [inline AK/StringFloatingPointConversions.cpp:15]: This link doesn't work because of the period, maybe we can put a space between the link and period.

**Comment by davidot** [comment]
Also in case you're interested there are even newer algorithms like dragonbox: https://github.com/jk-jeon/dragonbox
which is used by libfmt

File: AK/StringFloatingPointConversions.cpp (hunk 4 above)
Comment (Hendiadyoin1) [reply AK/StringFloatingPointConversions.cpp]: Its a struct array, so renaming is enough
`AK::Array` just has some runtime  bounds checks IIRC otherwise its just a normal c Array

File: AK/StringFloatingPointConversions.cpp (hunk 5 above)
Comment (Hendiadyoin1) [reply AK/StringFloatingPointConversions.cpp]: AK/FloatingPoint has this I think

File: AK/StringFloatingPointConversions.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (DanShaders) [reply AK/StringFloatingPointConversions.h]: I've used FloatingPointStringConversions.h as a base for this file (and it has the same omitted "it").
Maybe just completely delete pragma? For instance, AK/FloatingPoint.h does not say anything about kernel.

File: AK/StringFloatingPointConversions.h [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+///  2. For +0, it is {.sign = 0, .fraction = 0, .exponent = 0},
+///     for -0, is {.sign = 1, .fraction = 0, .exponent = 0},
+///     for +inf, -inf, and NaN is undefined.
+///  3. `(-1) ^ sign * fraction * 10 ^ exponent', computed with an infinite precision, rounds to
+///     `value' in the half to even mode.
+///  4. len(str(fraction)) is minimal.
+///  5. `abs((-1) ^ sign * fraction * 10 ^ exponent - value)' is minimal.
+///  6. fraction is minimal.
+template<typename FloatingPoint = double>
+FloatingPointExponentialForm convert_floating_point_to_fixed_decimal(FloatingPoint value);
```
[the remark this answers is not in the document]
Comment (DanShaders) [reply AK/StringFloatingPointConversions.h:37]: Not sure what I meant in the first place. Would "decimal_exponential_form" work?

File: AK/StringFloatingPointConversions.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (davidot) [reply AK/StringFloatingPointConversions.h]: I think this is more likely to (unintentionally) be used by the KERNEL when printing stuff, so I think keeping the pragma is good, perhaps only on the cpp file, so there are no implementations of it for KERNEL might be good enough

Anyway for now let's keep it, since this is still a WIP PR

File: AK/StringFloatingPointConversions.h [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+///  1. sign is either 0 or 1, fraction is a non-negative number, exponent is an integer.
+///  2. For +0, it is {.sign = 0, .fraction = 0, .exponent = 0},
+///     for -0, is {.sign = 1, .fraction = 0, .exponent = 0},
+///     for +inf, -inf, and NaN is undefined.
+///  3. `(-1) ^ sign * fraction * 10 ^ exponent', computed with an infinite precision, rounds to
+///     `value' in the half to even mode.
+///  4. len(str(fraction)) is minimal.
+///  5. `abs((-1) ^ sign * fraction * 10 ^ exponent - value)' is minimal.
+///  6. fraction is minimal.
+template<typename FloatingPoint = double>
```
[the remark this answers is not in the document]
Comment (DanShaders) [reply AK/StringFloatingPointConversions.h:36]: I did it like it was done in FloatingPointStringConversions.h
https://github.com/SerenityOS/serenity/blob/db2e1bfa027774ff1e1f883c4ddadc277c1e6e1e/AK/FloatingPointStringConversions.h#L62

File: AK/StringFloatingPointConversions.cpp [hunk 12]
```diff
@@ -0,0 +1,2278 @@
...
+
+    // for +inf, -inf, and NaN is undefined.
+    // To be on the safe side, return something meaningless rather than just not checking.
+    if (e == 2047) [[unlikely]] {
+        return { 0, 0, 0 };
+    }
+
+    // Step 1. Decode the floating point number, and unify normalized and subnormal cases.
+    u64 mf = (e == 0 ? 0 : (1ull << LEN_M)) + m;
+    i32 ef = (e == 0 ? 1 : e) - BIAS - LEN_M;
```
[the remark this answers is not in the document]
Comment (DanShaders) [reply AK/StringFloatingPointConversions.cpp]: These names come from the article, so in theory they should ease checking the code. (At least, it was definitely easier to write code with them)

**Comment by DanShaders** [comment]
Well, I haven't seen it when I was doing the initial research.
Anyway, I believe that what is currently implemented would already be a significant improvement over a for loop with fmod. So, I want this PR to stay with the Ryu and then potentially swap it for something better.

**Comment by davidot** [comment]
> Well, I hadn't seen it when I was doing the initial research. Anyway, I believe that what is currently implemented would already be a significant improvement over a for loop with fmod. So, I want this PR to stay with the Ryu and then potentially swap it for something better.

Yes we can just start with Ryu, don't think we need the performance difference of dragonbox or the others

**Comment by DanShaders** [comment]
Firstly, I've stress tested double conversion (checked 10^11 random bit patterns of double and all doubles with mantissa < 2^20 or mantissa > 2^52 - 2^20) and brute forced all possible floats against dragonbox. Secondly, I've run some benchmarks: our implementation of Ryu converts double in 30ns and float in 35ns on average on my machine (i7-12700h) (versus constant 10ns from Dragonbox).

All of the benchmarks are [here](https://github.com/DanShaders/serenity-arithmetic-benchmark). I do not really want to leave this code in the separate repository but do not know where to place it here (and it also depends on the external library).

I guess the PR is already too big for first, so will leave the benchmarking code intact as well as the other double -> string conversions in AK and LibJS.

**Comment by jk-jeon** [comment]
@DanShaders  The minmax Euclid algorithm appearing in the Adams' paper is not entirely correct. For example, with the Python code you embedded in the cpp file, `minmax_euclid(3,8,5)` prints `(1,2)` which means the minimum is `1` and the maximum is `6`. But `3 * 5 mod 8 = 7` so actually `7` is also possible. I actually do not believe the proof written in the paper. When I was looking at it, after several trials I just gave up comprehending it as it just looked very suspicious to me.

Also according to what I remember the algorithm as written is ridiculously slow. No idea how long did it actually take for giving you the table though. I guess maybe Adams himself is also aware of these issues and maybe the one that is currently in the reference implementation repo is probably the one with these issues addressed.

Even if the verification algorithm is wrong, the actual lookup table you got is probably okay. But who can be perfectly sure about that. I guess maybe you need to perform either the complete exhaustive test (it sounds like you already did for `float`, and it's practically impossible for `double`), or re-do the precision sufficiency check with a correct algorithm.

In [this paper](https://github.com/jk-jeon/Grisu-Exact/blob/master/other_files/Grisu-Exact.pdf), page 22, you can find an alternative algorithm which gives the exact min/max, which I believe to be actually correct. (I guess likely the one Adams is currently using is something similar to this.) Also this runs much faster than the one given in the Adams' paper. With my C++ implementation it runs instantaneously, though it can be somewhat slower in Python. I wrote a very long-winded mathematical proof of this algorithm, and quite confident about the algorithm's correctness, although nobody proofread the proof I wrote. Actually I do not use this algorithm anymore because it has been superseded by a more generic routine of computing so-called the best rational approximations. I would say this is indeed a strictly better approach, but if you want a quick drop-in replacement for the current implementation of `minmax_euclid`, then the one written in the paper mentioned above might be a good fit.

**Review by linusg — CHANGES_REQUESTED** [review]
Please squash all the fixup commits, this is much harder to review if you're changing a bunch of stuff cross commits until arriving at the final implementation that you want to get merged.

**Comment by DanShaders** [comment]
@linusg Done
@jk-jeon As for speed, it wasn't that bad, though I only tried running it for numbers up to 80 bits. With the "more likely to be correct algorithm" the bounds (B0 and B1) are the same, so the tables should indeed be correct.
Actually, I'm still not being able to get bounds mentioned in the Ryu paper (mine are off by 1 or 2). This is not a big deal for doubles, but for floats mine produce coefficients larger than UINT64_MAX, so I had to manually overwrite them and check correctness via brute force. In the reference implementation, there are some unclear bit length computations around [here](https://github.com/ulfjack/ryu/blob/75d5a85440ed356ad7b23e9e6002d71f62a6255c/src/main/java/info/adams/ryu/analysis/ComputeRequiredBitSizes.java#L77) where these numbers are coming from. I'm wondering if you've encountered the same problem before.

**Comment by jk-jeon** [comment]
@DanShaders First of all, I have never implemented Ryu by myself, although I did implement several different algorithms doing the same thing and utilized the (corrected version of) minmax Euclid for that.

The exact table entries I used are, if I recall correctly, somewhat different from Ryu. I aligned all of the first 1's to the MSB of the table entries, not sure if Ryu also does that. If not, that's probably the most prominent difference. (Plus the one in the reference repo has more entries than necessary, because it also contains the ones needed for the parsing.)

>With the "more likely to be correct algorithm" the bounds (B0 and B1) are the same, so the tables should indeed be correct.

Did you already check that? So fast😁

>This is not a big deal for doubles, but for floats mine produce coefficients larger than UINT64_MAX

From which exponent did you get that?

>In the reference implementation, there are some unclear bit length computations around [here](https://github.com/ulfjack/ryu/blob/75d5a85440ed356ad7b23e9e6002d71f62a6255c/src/main/java/info/adams/ryu/analysis/ComputeRequiredBitSizes.java#L77) where these numbers are coming from.

Isn't this just Lemma 3.4 in the Ryu paper? Could you elaborate why do you think it's unclear?

**Comment by DanShaders** [comment]
@jk-jeon Yes, I'm definitely not aligning table entries.

> Did you already check that? So fast😁

Basically I just skimmed through the proof, copied the code from the article and checked that $B_0$ and $B_1$ are the same.

> From which exponent did you get that?

Oh, technically I have a bug in my script (I forgot about the counterintuitive rounding towards zero of // for negative numbers in Python). Nonetheless, the paper also seems to have a problem:
I assume that formulas in subsection 3.4 step 3' case $e_2 < 0$ should be the following:

$$
\begin{align}
q &= \max\left(0, \lfloor -e_2 \log_{10} 5 \rfloor - 1 \right) \\
k &= \lceil \log_2 5^{-e_2 - q} \rceil - B_1 \\
\end{align}
$$

For $e_2 = -140$ and $B_1 = 63$, we get $q = 96$ and $k = 40$. However, using lemma 3.4 we get

$$
k \le \log_2 \left( \frac{\min((w \cdot 5^{-e2-q})\mod 2^q)}{\max(w)} \right) < 39.926
$$

assuming $\max(w) = 67'108'862$ and substituting $w = 45'038'809$ for minimum. So, that is why 63 is a _supposedly_ incorrect bound.

However, if I round everything correctly, using 64 instead of 63 is not a problem. As everything works regardless, I'm not really certain but I might fix the rounding in my code.

> Could you elaborate why do you think it's unclear?

I was not able to understand how those bit length manipulations resulted in the correct rounding, and, as we can see, they didn't.

**Comment by jk-jeon** [comment]
@DanShaders
>Basically I just skimmed through the proof, copied the code from the article and checked that $B_{0}$ and $B_{1}$ are the same.

Actually I'm quite doubtful about that. Are you sure you get more than 64-bits for IEEE-754 binary32?

>Oh, technically I have a bug in my script (I forgot about the counterintuitive rounding towards zero of // for negative numbers in Python)

You mean so your $B_{0}$ and $B_{1}$ are actually not more that 64 for IEEE-754 binary32? Also so sad that C++ also does the same thing for integer division... Apparently it is because that's how most of the actual hardware work. ([Here](https://cplusplus.github.io/CWG/issues/614.html) is the defect report which mandated this behavior.)

>Nonetheless, the paper also seems to have a problem:

As far as I can tell there is nothing wrong in the statement/proof of Lemma 3.4. I guess you mean that $q$ is not equal to what it is supposed to be, which kind of makes sense b/c it doesn't match the $q$ given in Lemma 3.2. Probably the author mistakenly thought that $\log_{10}5$ is equal to $\log_{5}2$. I don't know if the specific $q$ you proposed is the right one though, as it's still off by 1 from what's given in Lemma 3.2. The proof of $e_{2}<0$ case of Lemma 3.2 is omitted in the paper, so maybe, the author is again wrong at this point, but I didn't check that. Are you sure $-1$ is the correct one?

>However, if I round everything correctly, using 64 instead of 63 is not a problem.

As far as everything is inside $64$-bits there should be no problem yeah.

**Comment by DanShaders** [comment]
@jk-jeon

> Actually I'm quite doubtful about that. Are you sure you get more than 64-bits for IEEE-754 binary32?

No, I'm not anymore.

> You mean so your $B_0$ and $B_1$ are actually not more that 64 for IEEE-754 binary32?

Exactly.

> Also so sad that C++ also does the same thing for integer division...

Off topic: I have CP background where we usually didn't dealt with negative numbers in the context of floor/ceil (except for some problems on number theory), so I'm very used to floor(a/b) = a // b)


> As far as I can tell there is nothing wrong in the statement/proof of Lemma 3.4. I guess you mean that q is not equal to what it is supposed to be, which kind of makes sense b/c it doesn't match the q given in Lemma 3.2. Probably the author mistakenly thought that log10⁡5 is equal to log5⁡2. I don't know if the specific q you proposed is the right one though, as it's still off by 1 from what's given in Lemma 3.2. The proof of e2<0 case of Lemma 3.2 is omitted in the paper, so maybe, the author is again wrong at this point, but I didn't check that.

I don't see how the proof depends on the exact value of q, so it should work at least for $q \in [0; -\lfloor e_2 \log_{10} 5\rfloor\]$. The contradiction is that $k$, given by $k = \lceil \log_2 5^{-e_2 - q} \rceil - B_1$ with $B_1=63$, should be allowed ("and then use B1 to determine a legal value for k as") but it is not as $40 > 39.926$. Of course, I'm not entirely sure about this exact formula for $q$ but it is what I use (and successfully stress against dragonbox) and it follows the logic of "the largest which does not break loop invariant - 1".

Just to clarify: I define $B_1$ as the minimal number such that $\forall -e2 - q : \lceil \log_25 * (-e2 - q) \rceil - B_1 <= k_\text{max}$ and then use "the bit length of the coefficients is no more than $B_1$" as the corollary.

**Comment by jk-jeon** [comment]
@DanShaders There is no contradiction if you choose $q = \lfloor-e_{2}\log_{10}5\rfloor$. Then $e_{2}=-140$, $q=97$, so according to your definition, $k$ becomes $37$. I haven't checked what is the bound Lemma 3.4 gives in this case, but probably it's more than $37$.

**Comment by DanShaders** [comment]
@jk-jeon The value is somewhere between 39 and 40.
As far as I can tell, they are using two different formulas for q: one with -1 (https://github.com/ulfjack/ryu/blob/75d5a85440ed356ad7b23e9e6002d71f62a6255c/ryu/d2s.c#L158) and one without (https://github.com/ulfjack/ryu/blob/75d5a85440ed356ad7b23e9e6002d71f62a6255c/ryu/f2s.c#L112)...

**Comment by DanShaders** [comment]
@jk-jeon Hmm, earlier d2s also used formula without -1 if NICER_OUTPUT was not defined. https://github.com/ulfjack/ryu/blob/365dd77ddfca1e38ae88099b53d4c99c21b82a0e/ryu/d2s.c#L525

**Comment by jk-jeon** [comment]
>The value is somewhere between 39 and 40.

I just ran a computation. The bound is between 39 and 40 when $q = \lfloor -e_{2}\log_{10}5\rfloor - 1$. If you use $q = \lfloor -e_{2}\log_{10}5\rfloor$ instead, then it is between 43 and 44.

>Hmm, earlier d2s also used formula without -1 if NICER_OUTPUT was not defined.

My conspiracy theory: the formula without -1 is the correct one, but later they discovered that the one with -1 actually leads to the same output for `double` and has a bunch of benefits so they just got stuck to it.

**Comment by DanShaders** [comment]
> My conspiracy theory: the formula without -1 is the correct one, but later they discovered that the one with -1 actually leads to the same output for `double` and has a bunch of benefits so they just got stuck to it.

This seems reasonable.

I guess we are not discussing the PR anymore. Just to make sure that we are on the same page: would you say that the code in terms of the algorithms used should be correct?

**Comment by jk-jeon** [comment]
@DanShaders As I pointed out in the first comment I actually haven't implemented Ryu by myself, in fact I even have never read the Ryu paper thoroughly except for some parts from Section 3.2. So I cannot have any idea on your actual code on Ryu itself. (I can certainly believe that it is correct as you did a nice stress test against Dragonbox reference impl.) The only thing I was worried about was the correctness of `minmax_euclid`. I have seen several people who reimplemented Ryu but I am not aware of anyone else who actually re-created the table by directly using `minmax_euclid`. Since `minmax_euclid` has some issue and my impression was that Adams is actually using a different routine, I just wanted to make sure that the issue is not really materialized. As it seems like you replaced the embedded Python code into the one using the corrected algorithm, I'm all happy now.

File: AK/FloatingPoint.h [hunk 13]
```diff
@@ -12,6 +12,57 @@
 
 namespace AK {
 
+template<typename T>
+union FloatExtractor;
+
+#if ARCH(I386) || ARCH(X86_64) || ARCH(AARCH64)
+// This assumes long double is 80 bits, which is true with GCC on Intel platforms
```
Comment (davidot) [inline AK/FloatingPoint.h:19]: I know it didn't have this before but maybe we can add a `static_assert(sizeof(long double) == 10)` here to validate what this comment is saying?

File: AK/FloatingPoint.h [hunk 14]
```diff
@@ -182,6 +233,8 @@ constexpr float convert_to_native_float(I input) { return float_to_float<SingleF
 
 }
 
+using AK::FloatExtractor;
+
```
Comment (davidot) [inline AK/FloatingPoint.h:237]: I don't think we usually have empty lines between these lines

File: AK/StringFloatingPointConversions.cpp [hunk 15]
```diff
@@ -0,0 +1,1126 @@
...
+            return (b >> static_cast<u32>(shift - 64)).low();
+        }
+    };
+
+    return inner_convert_floating_point_to_decimal_exponential_form(value, multiply_and_shift);
+}
+
+// Step 0. Precompute lookup tables for the given floating point type.
+// Lookup tables was generated using the following Python script.
+/*
```
Comment (davidot) [inline AK/StringFloatingPointConversions.cpp:227]: I'm not sure if we want to keep this entire code in here as a comment.
But I can't think of a better way to do this so unless someone has a nice idea we can just leave it.

File: AK/StringFloatingPointConversions.cpp [hunk 16]
```diff
@@ -0,0 +1,1126 @@
...
+    return u128 { low, high };
+#else
+    return u128 { a }.wide_multiply(u128 { b }).low;
+#endif
+}
+
+template<>
+FloatingPointExponentialForm convert_floating_point_to_decimal_exponential_form<float>(float value)
+{
+    auto multiply_and_shift = [&](u64 operand, u64 multiplier, i32 shift) {
```
Comment (davidot) [inline AK/StringFloatingPointConversions.cpp:189]: Since this lambda doesn't use `value` you can just use an empty capture group `[](u64 operand, ...`

File: AK/StringFloatingPointConversions.cpp [hunk 17]
```diff
@@ -0,0 +1,1126 @@
...
+            return static_cast<u64>(result >> static_cast<u32>(shift));
+    };
+
+    return inner_convert_floating_point_to_decimal_exponential_form(value, multiply_and_shift);
+}
+
+template<>
+FloatingPointExponentialForm convert_floating_point_to_decimal_exponential_form<double>(double value)
+{
+    auto multiply_and_shift = [&](u64 operand, u64 const multiplier[2], i32 shift) {
```
Comment (davidot) [inline AK/StringFloatingPointConversions.cpp:207]: Same for this one

File: AK/StringFloatingPointConversions.cpp [hunk 18]
```diff
@@ -0,0 +1,1126 @@
...
+    if (exponent == Extractor::exponent_max) [[unlikely]]
+        return { 0, 0, 0 };
+
+    // Step 1. Decode the floating point number, and unify normalized and subnormal cases.
+    u64 real_mantissa = (exponent == 0 ? 0 : (1ull << Extractor::mantissa_bits)) + mantissa;
+    i32 real_exponent = (exponent == 0 ? 1 : exponent) - Extractor::exponent_bias - Extractor::mantissa_bits;
+    // abs(value) = real_mantissa * 2 ^ real_exponent
+
+    // Step 2. Determine the interval of information-preserving outputs.
+    // u, v, w are, respectively, lower bound for answer, exact value and upper bound for answer.
```
Comment (davidot) [inline AK/StringFloatingPointConversions.cpp:72]: I still think that if `u`, `v` and `w` have a more meaning with bounds they should have those names but with the comments of the algorithm added this is already more understandable. 
So you don't have to change this.

File: AK/StringFloatingPointConversions.h (hunk 11 above)
[the remark this answers is not in the document]
Comment (davidot) [reply AK/StringFloatingPointConversions.h:37]: Can we limit the template with the FloatingPoint concept?
`template<FloatingPoint T>`? (You'd have to change the parameter as well)

File: Tests/AK/TestStringFloatingPointConversions.cpp [hunk 19]
```diff
@@ -0,0 +1,66 @@
...
+            convert_floating_point_to_decimal_exponential_form(static_cast<double>(value)), \
+            (AK::FloatingPointExponentialForm { sign, fraction, exponent }));               \
+    } while (false)
+
+// Tests here only check basic cases. While writing, I mostly relied on the benchmarks and
+// stress tests, which can be found at https://github.com/DanShaders/serenity-arithmetic-benchmark/blob/master/StringFloatingPointConversions.cpp .
+
+TEST_CASE(double_conversion)
+{
+    DOES_CONVERT_DOUBLE_TO(0, 0, 0, 0);
```
Comment (davidot) [inline Tests/AK/TestStringFloatingPointConversions.cpp:36]: I guess this will be excercised in the rest of the system but I would still like a couple of more trivial tests like: `0.1` `0.2` `0.3` as those as "famous" examples with `0.1 + 0.2 !== 0.3` in javascript.

Also at least the negative zero edge case.

File: AK/StringFloatingPointConversions.cpp [hunk 20]
```diff
@@ -0,0 +1,1126 @@
...
+
+    // For +0, it is {.sign = 0, fraction = 0, exponent = 0},
+    // for -0, is {.sign = 1, fraction = 0, exponent = 0},
+    if (exponent == 0 && mantissa == 0)
+        return { sign, 0, 0 };
+
+    // for +inf, -inf, and NaN is undefined.
+    // To be on the safe side, return something meaningless rather than just not checking.
+    if (exponent == Extractor::exponent_max) [[unlikely]]
+        return { 0, 0, 0 };
```
Comment (davidot) [inline AK/StringFloatingPointConversions.cpp:64]: Would it be better to make this a `VERIFY`? 
This would indicate to the programmer that you should filter those values out first.
Currently this would just give the same results as `0` which might make it hard to debug

**Review by davidot — CHANGES_REQUESTED** [review]
Nice work once again!

I have a couple of small things but some of them are really comments not necessarily things that must be fixed. I think with some more text from the paper the code is already more readable.

File: AK/FloatingPoint.h (hunk 13 above)
Comment (BertalanD) [reply AK/FloatingPoint.h:19]: FYI, `long double` is the same as `double` (64 bits) in the Apple arm64 ABI.

**Comment by ADKaster** [comment]
Looks like I created a conflict by merging https://github.com/SerenityOS/serenity/pull/15757, sorry about that!

File: AK/FloatingPoint.h (hunk 13 above)
Comment (ADKaster) [reply AK/FloatingPoint.h:19]: sizeof(long double) is 12 in -m32 mode with both gcc and clang. so seems an x86 32-bitism. on 64-bit both compilers report 16.

https://godbolt.org/z/Ps7TWsj3j

Comment (ADKaster) [reply AK/FloatingPoint.h:19]: So any assumptions about IEEE 754 quad floats vs Intel 8087 "Extended Precision" 80-bit floats should probably be verified by static_asserts, yes. As-is, the comment is incorrect in the face of the preprocessor `#if`

Comment (DanShaders) [reply AK/FloatingPoint.h]: It seems that `sizeof` and `bit_cast<long double>` do not work, so I ended up checking for the value of `__LDBL_MAX__`. (Yes, I'm aware of NumericLimits but don't think it is rational to include it for one compile-time check)

**Comment by DanShaders** [comment]
I'm kinda losing hope for getting this merged. It was supposed to be a quick fix :)

File: AK/FloatingPoint.h (hunk 13 above)
Comment (ADKaster) [reply AK/FloatingPoint.h]: Eh. It will only be 80-bit long double on i686 and such, or x86_64 platforms where you used -mlong-double-80. Otherwise it will "probably" be 16 bit IEEE 754 quad-float. ... Except on apple platforms like Daniel mentioned, where double and long double are the same thing. 

So I think to resolve this we should update the check by removing the current static assert, and updating the comment to say "FIXME: assumes long double is Intel x87 80-bit floating point, which is only true on some x86 platforms" and leave it for later

Comment (ADKaster) [reply AK/FloatingPoint.h]: We shouldn't be using long double in any serenity code at the moment other than the userspace emulator, so this being not quite right will only affect ports.

Comment (DanShaders) [reply AK/FloatingPoint.h]: Really? https://godbolt.org/z/T8Kx1fG6r

Comment (DanShaders) [reply AK/FloatingPoint.h]: Since we are not targeting MSVC, we can use __LDBL_MANT_DIG__ and __LDBL_MAX_EXP__ to be absolutely sure.
https://godbolt.org/z/7TxrYbqWx

**Review by davidot — APPROVED** [review]
I think with the static assert I'm fine with merging this.
We can always iterate on it later and adding the other double printing code can just happen in a follow up PR.
(Note I didn't look at the code again as I don't have much time currently)

**Comment by Vbitz** [comment]
This merge breaks the build on M1 Mac. macOS Version 12.6

```
In file included from /Users/joshua/dev/org/serenity/AK/StringFloatingPointConversions.cpp:9:
/Users/joshua/dev/org/serenity/Meta/Lagom/../../AK/FloatingPoint.h:21:1: error: floating constant exceeds range of 'long double' [-Werror=overflow]
   21 | static_assert(__LDBL_MAX__ == 1.189731495357231765e4932L);
      | ^~~~~~~~~~~~~
```

**Comment by ADKaster** [comment]
Yep see #15920 for the current attempted fix.
